### PR TITLE
[doc] start of best practices section in contributing.rst

### PR DESCRIPTION
**Pull request by timotheecour**
/cc @Araq @mratsim @narimiran @dom96 

I always wished there was a "best practices" section; let's start one and grow it over time, keeping it up to date with language improvements. It saves times for both PR authors and reviewers, we can just refer to this doc in code reviews, using anchors to the specific paragraph. 

There are plenty of posts on forum requesting such best practices section:

* [Lessons learned/idioms - Nim forum](https://forum.nim-lang.org/t/4124#25693)
* https://forum.nim-lang.org/t/4099#25496 "How should I handle testing these examples? ..."
* https://forum.nim-lang.org/t/1978#13958 "what are the best practices?"
* [Zen for NIM? - Nim forum](https://forum.nim-lang.org/t/1377#8593) "What are the recommended Practices between soo many choices? ..."
* https://forum.nim-lang.org/t/871#5177 "And are there other best practices in Nim for adding..."
* https://forum.nim-lang.org/t/608#3347 "As with the rest of Nim, some more docs and 'best practices' would be helpful."

## note
* I don't mind other guides containing such best practices (eg "external links" below), but it's desirable to have such a "best practices" section in Nim repo, version controlled along the rest of doc/code, so that it's always in sync with compiler/language, are updated when newer idioms become preferred, and agreed upon for code reviews (aka: no endless debates in each indivudual PR)
* if a rule is too controversial, let's skip it (or mark it as "controversial")

## external links
* https://scripter.co/notes/nim
* Nim in Action
* https://github.com/status-im
* https://github.com/nim-lang/Nim/wiki/Style-Guide-for-Nim-Code
* https://nim-lang.org/docs/nep1.html seems to cover mostly style (eg whitespace, naming) rather than semantics. This "best practices" is not intended to duplicate what's in there.  One possibility is to move the paragraph "Coding Conventions" to this "best practices" page so that the split is even clearer (style vs semantics)

### CODE REVIEW

File: doc/contributing.rst [hunk 1]
```diff
@@ -248,6 +248,94 @@ or
...
+
+  quit() # bad in almost all cases
+  doAssert() # preferred
+
+.. _tests_use_doAssert:
+Use ``doAssert`` (or ``require``, etc), not ``assert`` in all tests.
+
+.. code-block:: nim
+
+  runnableExamples: assert foo()
```
Comment (kaushalmodi) [inline doc/contributing.rst]: Did you forget to add `# bad` and `# preferred` comments here?

Comment (timotheecour) [reply doc/contributing.rst]: @kaushalmodi done; 
maybe you can use this: [github's new 'Suggested change' in code review is great; should be used - Nim forum](https://forum.nim-lang.org/t/4317)
I'm gonna start using it whenever possible in code reviews

Comment (kaushalmodi) [reply doc/contributing.rst]: @timotheecour Yes, I saw that. It's still not clear how to use that.. I'll read up on that before using it. Thanks :)

File: doc/contributing.rst [hunk 2]
```diff
@@ -248,6 +248,94 @@ or
...
+  doAssert isValid() == true
+  doAssert isValid() # preferred
+
+.. _noimplicitbool:
+Reduce parenthesis nesting
+
+.. code-block:: nim
+
+  doAssert (not 1 == 2)
+  doAssert: not 1 == 2  # preferred
```
Comment (Araq) [inline doc/contributing.rst]: No, this is not the preferred variant!

File: doc/contributing.rst [hunk 3]
```diff
@@ -248,6 +248,94 @@ or
 
 the first is preferred.
 
+Best practices
+=============
+
+Note: these are gneral guidelines, not hard rules; there are always exceptions.
```
Comment (Araq) [inline doc/contributing.rst]: typo.

File: doc/contributing.rst [hunk 4]
```diff
@@ -248,6 +248,94 @@ or
...
+Immediately invoked lambdas (https://en.wikipedia.org/wiki/Immediately-invoked_function_expression)
+
+.. code-block:: nim
+
+  let a = (proc (): auto = getFoo())()
+  let a = block:  # preferred
+    getFoo()
+
+.. _design_for_ufcs:
+Design with UFCS (method call syntax) chaining in mind
```
Comment (Araq) [inline doc/contributing.rst]: Nim does not use the term "UFCS". Anywhere.

File: doc/contributing.rst [hunk 5]
```diff
@@ -248,6 +248,94 @@ or
...
+Delegate printing to caller: return ``string`` instead of calling ``echo``
+rationale: it's more flexible (eg allows caller to call custom printing,
+including prepending location info, writing to log files, etc).
+
+.. code-block:: nim
+
+  proc foo() = echo "bar" # bad
+  proc foo(): string = "bar" # preferred (usually)
+
+.. _use_Option:
```
Comment (Araq) [inline doc/contributing.rst]: That's still an ongoing debate. :P

File: doc/contributing.rst (hunk 4 above)
Comment (kaushalmodi) [reply doc/contributing.rst]: I wish it did, I think that will improve the exposure of this awesome feature. May be, mention it as something like "method call syntax, which is similar to UFCS in few other languages"?

File: doc/contributing.rst (hunk 5 above)
Comment (timotheecour) [reply doc/contributing.rst]: marked it as such

File: doc/contributing.rst (hunk 2 above)
Comment (timotheecour) [reply doc/contributing.rst]: removed this section

File: doc/contributing.rst (hunk 3 above)
Comment (timotheecour) [reply doc/contributing.rst]: fixed

File: doc/contributing.rst (hunk 4 above)
Comment (timotheecour) [reply doc/contributing.rst]: reworded as suggested

**Comment by timotheecour** [comment]
PTAL

**Comment by LemonBoy** [comment]
Pro-tip: since you're working on something that doesn't really need a CI run every time you can add `[ci skip]` to your commit message titles.
This way you don't clog the pipeline with useless jobs that may be needed by other PRs.

This convention is supported by [Appveyor](https://www.appveyor.com/docs/how-to/filtering-commits/#skip-directive-in-commit-message) and [Travis](https://docs.travis-ci.com/user/customizing-the-build/#skipping-a-build).

**Comment by narimiran** [comment]
> One possibility is to move the paragraph "Coding Conventions" [from NEP-1] to this "best practices" page so that the split is even clearer (style vs semantics)

What about making this (best style practices) NEP-2, so it both more visible and can be shared more easily (IRC, Telegram, etc.), and then `contributing.rst` can just have a link to it?

**Comment by timotheecour** [comment]
@narimiran
> What about making this (best style practices) NEP-2

no objection, feel free to submit PR for that (maybe, with the paragraph "Coding Conventions" moved as discussed above)

@LemonBoy 
> Pro-tip: since you're working on something that doesn't really need a CI run every time you can add [ci skip] to your commit message titles.

Wow thanks for the tip! I will add it in "the git stuff" in subsequent PR

**Pull request merged**